src/sudoku.hs

```diff
@@ -0,0 +1,136 @@
+module Sudoku where
+import GHC.Arr (fill)
+import Data.List
+
+-- | A representation of the values in a sudoku from 0-9 TODO: SHOULD BE RENAMED!!!!!
+data SudVal = One | Two | Three | Four | Five | Six | Seven | Eight | Nine
+    deriving (Eq, Enum, Bounded)
+
+instance Show SudVal where
+    show One = "1"
+    show Two = "2"
+    show Three = "3"
+    show Four = "4"
+    show Five = "5"
+    show Six = "6"
+    show Seven = "7"
+    show Eight = "8"
+    show Nine = "9"
+
+data Line = Horizontal | Vertical
+    deriving (Eq)
+
+instance Show Line where
+    show Horizontal = "-"
+    show Vertical = "|"
+
+data Note = Candidate SudVal | Line Line SudVal
+    deriving (Eq)
+
+instance Show Note where
+    show (Candidate x) = show x
+    show (Line l x) = show l ++ show x
+
+data Value
+    = Filled SudVal
+    | Note [Note]
+    | Empty
+    deriving (Eq)
```

Comment:
* I would like to see comments that explain the logical meaning of the note annotations.
* It's better to use separate fields for different types of annotations instead of putting them all into the same list. Then extracting and modifying them becomes much easier.